File: predexonooor/predexon.py
from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Any, Literal

import requests


class PredexonApiError(RuntimeError):
    def __init__(self, *, status_code: int, error: str | None, message: str | None, request_id: str | None):
        super().__init__(f"Predexon API error {status_code}: {error} {message} ({request_id})")
        self.status_code = status_code
        self.error = error
        self.message = message
        self.request_id = request_id


def _raise_for_error(resp: requests.Response) -> None:
    if 200 <= resp.status_code < 300:
        return
    try:
        payload = resp.json()
    except Exception:
        payload = {}
    raise PredexonApiError(
        status_code=resp.status_code,
        error=payload.get("error"),
        message=payload.get("message"),
        request_id=payload.get("requestId") or resp.headers.get("x-request-id"),
    )


@dataclass(frozen=True)
class DataClient:
    session: requests.Session
    base_url: str = "https://api.predexon.com"
    timeout_seconds: float = 20.0

    def health(self) -> dict[str, Any]:
        r = self.session.get(f"{self.base_url}/health", timeout=self.timeout_seconds)
        _raise_for_error(r)
        return r.json()

    def list_polymarket_markets(self, *, status: str = "open", sort: str = "volume", limit: int = 30) -> dict[str, Any]:
        r = self.session.get(
            f"{self.base_url}/v2/polymarket/markets",
            params={"status": status, "sort": sort, "limit": limit},
            timeout=self.timeout_seconds,
        )
        _raise_for_error(r)
        return r.json()

    def get_outcome(self, *, predexon_id: str, routable_only: bool = True) -> dict[str, Any]:
        r = self.session.get(
            f"{self.base_url}/v2/outcomes/{predexon_id}",
            params={"routable_only": str(routable_only).lower()},
            timeout=self.timeout_seconds,
        )
        _raise_for_error(r)
        return r.json()

    def polymarket_orderbooks(self, *, token_id: str, start_ms: int, end_ms: int, limit: int = 1) -> dict[str, Any]:
        r = self.session.get(
            f"{self.base_url}/v2/polymarket/orderbooks",
            params={"token_id": token_id, "start_time": start_ms, "end_time": end_ms, "limit": limit},
            timeout=self.timeout_seconds,
        )
        _raise_for_error(r)
        return r.json()

    def limitless_orderbooks(self, *, market_slug: str, start_ms: int, end_ms: int, limit: int = 1) -> dict[str, Any]:
        r = self.session.get(
            f"{self.base_url}/v2/limitless/orderbooks",
            params={"market_slug": market_slug, "start_time": start_ms, "end_time": end_ms, "limit": limit},
            timeout=self.timeout_seconds,
        )
        _raise_for_error(r)
        return r.json()


Venue = Literal["polymarket", "limitless", "predict", "opinion", "hyperliquid"]


@dataclass(frozen=True)
class TradeClient:
    session: requests.Session
    base_url: str = "https://trade.predexon.com"
    timeout_seconds: float = 20.0

    def health(self) -> dict[str, Any]:
        r = self.session.get(f"{self.base_url}/health", timeout=self.timeout_seconds)
        _raise_for_error(r)
        return r.json()

    def create_account(self) -> dict[str, Any]:
        r = self.session.post(f"{self.base_url}/api/accounts/create", timeout=self.timeout_seconds)
        _raise_for_error(r)
        return r.json()

    def list_accounts(self) -> dict[str, Any]:
        r = self.session.get(f"{self.base_url}/api/accounts", timeout=self.timeout_seconds)
        _raise_for_error(r)
        return r.json()

    def get_account(self, *, account_id: str) -> dict[str, Any]:
        r = self.session.get(f"{self.base_url}/api/accounts/{account_id}", timeout=self.timeout_seconds)
        _raise_for_error(r)
        return r.json()

    def enable_venue(self, *, account_id: str, venue: Venue) -> dict[str, Any]:
        r = self.session.post(
            f"{self.base_url}/api/accounts/{account_id}/enable",
            json={"venue": venue},
            timeout=self.timeout_seconds,
        )
        _raise_for_error(r)
        return r.json()

    def get_balance(self, *, account_id: str, aggregated: bool = False) -> dict[str, Any]:
        r = self.session.get(
            f"{self.base_url}/api/accounts/{account_id}/balance",
            params={"aggregated": str(aggregated).lower()},
            timeout=self.timeout_seconds,
        )
        _raise_for_error(r)
        return r.json()

    def get_positions(self, *, account_id: str, aggregated: bool = False) -> dict[str, Any]:
        r = self.session.get(
            f"{self.base_url}/api/accounts/{account_id}/positions",
            params={"aggregated": str(aggregated).lower()},
            timeout=self.timeout_seconds,
        )
        _raise_for_error(r)
        return r.json()

    def place_order(
        self,
        *,
        account_id: str,
        venue: Venue,
        token_id: str,
        side: Literal["buy", "sell"],
        order_type: Literal["market", "limit"],
        size: float,
        price: float | None = None,
        client_id: str | None = None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "venue": venue,
            "market": {"tokenId": token_id},
            "side": side,
            "type": order_type,
            "size": str(size),
        }
        if price is not None:
            payload["price"] = str(price)
        if client_id:
            payload["clientId"] = client_id

        r = self.session.post(
            f"{self.base_url}/api/accounts/{account_id}/orders",
            json=payload,
            timeout=self.timeout_seconds,
        )
        _raise_for_error(r)
        return r.json()

    def router_place_order(
        self,
        *,
        account_id: str,
        predexon_id: str,
        side: Literal["buy", "sell"],
        order_type: Literal["market", "limit"],
        amount: float | None = None,
        size: float | None = None,
        price: float | None = None,
        bridge_enabled: bool | None = None,
        client_order_id: str | None = None,
        explain: bool = False,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {"predexonId": predexon_id, "side": side, "type": order_type}
        if amount is not None:
            payload["amount"] = amount
        if size is not None:
            payload["size"] = size
        if price is not None:
            payload["price"] = price
        if bridge_enabled is not None:
            payload["bridgeEnabled"] = bool(bridge_enabled)
        if client_order_id:
            payload["clientOrderId"] = client_order_id

        r = self.session.post(
            f"{self.base_url}/api/accounts/{account_id}/router/orders",
            params={"explain": str(explain).lower()},
            json=payload,
            timeout=self.timeout_seconds,
        )
        _raise_for_error(r)
        body = r.json()
        if isinstance(body, dict) and body.get("status") == "failed":
            return body
        return body


def now_ms() -> int:
    return int(time.time() * 1000)

File: src/predexon.ts
export type Venue = "polymarket" | "limitless";

export class PredexonApiError extends Error {
  statusCode: number;
  errorCode?: string;
  requestId?: string;

  constructor(opts: { statusCode: number; errorCode?: string; message?: string; requestId?: string }) {
    super(`Predexon API error ${opts.statusCode}: ${opts.errorCode ?? ""} ${opts.message ?? ""} (${opts.requestId ?? ""})`);
    this.statusCode = opts.statusCode;
    this.errorCode = opts.errorCode;
    this.requestId = opts.requestId;
  }
}

const pickApiKey = (url: string): string => {
  const host = new URL(url).host;
  const fallback = String(process.env.PREDEXON_API_KEY ?? "").trim();

  if (host === "api.predexon.com") {
    return String(process.env.PREDEXON_DATA_API_KEY ?? fallback).trim();
  }

  if (host === "trade.predexon.com") {
    return String(process.env.PREDEXON_TRADING_API_KEY ?? fallback).trim();
  }

  return fallback;
};

const withApiKey = (url: string, headers?: HeadersInit): HeadersInit => {
  const k = pickApiKey(url);
  if (!k) throw new Error("Missing Predexon API key (set PREDEXON_DATA_API_KEY / PREDEXON_TRADING_API_KEY)");
  return { ...(headers ?? {}), "x-api-key": k };
};

const parseJsonSafe = async (res: Response): Promise<any> => {
  try {
    return await res.json();
  } catch {
    return {};
  }
};

const request = async (url: string, init: RequestInit & { timeoutMs?: number } = {}): Promise<any> => {
  const timeoutMs = init.timeoutMs ?? 20_000;
  const controller = new AbortController();
  const t = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetch(url, { ...init, signal: controller.signal, headers: withApiKey(url, init.headers) });
    if (res.ok) return await res.json();
    const payload = await parseJsonSafe(res);
    throw new PredexonApiError({
      statusCode: res.status,
      errorCode: payload?.error,
      message: payload?.message,
      requestId: payload?.requestId ?? res.headers.get("x-request-id") ?? undefined
    });
  } finally {
    clearTimeout(t);
  }
};

export class DataClient {
  baseUrl = "https://api.predexon.com";

  health(): Promise<any> {
    return this.listPolymarketMarkets({ limit: 1 });
  }

  listPolymarketMarkets(params: { status?: string; sort?: string; limit?: number }): Promise<any> {
    const u = new URL(`${this.baseUrl}/v2/polymarket/markets`);
    u.searchParams.set("status", params.status ?? "open");
    u.searchParams.set("sort", params.sort ?? "volume");
    u.searchParams.set("limit", String(params.limit ?? 30));
    return request(u.toString());
  }

  getOutcome(predexonId: string, routableOnly = true): Promise<any> {
    const u = new URL(`${this.baseUrl}/v2/outcomes/${predexonId}`);
    u.searchParams.set("routable_only", String(routableOnly));
    return request(u.toString());
  }

  polymarketOrderbooks(params: { tokenId: string; startMs: number; endMs: number; limit?: number }): Promise<any> {
    const u = new URL(`${this.baseUrl}/v2/polymarket/orderbooks`);
    u.searchParams.set("token_id", params.tokenId);
    u.searchParams.set("start_time", String(params.startMs));
    u.searchParams.set("end_time", String(params.endMs));
    u.searchParams.set("limit", String(params.limit ?? 1));
    return request(u.toString());
  }

  limitlessOrderbooks(params: { marketSlug: string; startMs: number; endMs: number; limit?: number }): Promise<any> {
    const u = new URL(`${this.baseUrl}/v2/limitless/orderbooks`);
    u.searchParams.set("market_slug", params.marketSlug);
    u.searchParams.set("start_time", String(params.startMs));
    u.searchParams.set("end_time", String(params.endMs));
    u.searchParams.set("limit", String(params.limit ?? 1));
    return request(u.toString());
  }
}

export class TradeClient {
  baseUrl = "https://trade.predexon.com";

  health(): Promise<any> {
    return request(`${this.baseUrl}/health`);
  }

  createAccount(): Promise<any> {
    return request(`${this.baseUrl}/api/accounts/create`, { method: "POST" });
  }

  listAccounts(): Promise<any> {
    return request(`${this.baseUrl}/api/accounts`);
  }

  getAccount(accountId: string): Promise<any> {
    return request(`${this.baseUrl}/api/accounts/${accountId}`);
  }

  enableVenue(accountId: string, venue: Venue): Promise<any> {
    return request(`${this.baseUrl}/api/accounts/${accountId}/enable`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ venue })
    });
  }

  getBalance(accountId: string, aggregated = false): Promise<any> {
    const u = new URL(`${this.baseUrl}/api/accounts/${accountId}/balance`);
    u.searchParams.set("aggregated", String(aggregated));
    return request(u.toString());
  }

  getPositions(accountId: string, aggregated = false): Promise<any> {
    const u = new URL(`${this.baseUrl}/api/accounts/${accountId}/positions`);
    u.searchParams.set("aggregated", String(aggregated));
    return request(u.toString());
  }

  getDepositInfo(accountId: string): Promise<any> {
    return request(`${this.baseUrl}/api/accounts/${accountId}/deposit-info`);
  }

  placeOrder(params: {
    accountId: string;
    venue: Venue;
    predexonId?: string;
    market?: { tokenId?: string; marketSlug?: string };
    side: "buy" | "sell";
    type: "market" | "limit";
    amount?: number;
    size?: number;
    price?: number;
    clientId?: string;
  }): Promise<any> {
    const payload: any = {
      venue: params.venue,
      side: params.side,
      type: params.type
    };
    if (params.predexonId) payload.predexonId = params.predexonId;
    if (params.market) payload.market = params.market;
    if (params.amount !== undefined) payload.amount = String(params.amount);
    if (params.size !== undefined) payload.size = String(params.size);
    if (params.price !== undefined) payload.price = String(params.price);
    if (params.clientId) payload.clientId = params.clientId;

    return request(`${this.baseUrl}/api/accounts/${params.accountId}/orders`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(payload)
    });
  }
}

export const nowMs = (): number => Date.now();
